### PR TITLE
Implement a fundamental OGM framework.

### PR DESCRIPTION
**Pull request by supercmmetry**

### CODE REVIEW

File: core/domain.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package lucy
+
+type DomainType uint
+
+const (
+	Where     DomainType = 0
```
Comment (L04DB4L4NC3R) [inline core/domain.go]: Use enums here `iota`

File: core/lucy.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	Stop()
+	Sync() error
+}
+
+type KeyValuePair struct {
+	Key   string
+	Value interface{}
+}
+
+type Lucy struct {
```
Comment (L04DB4L4NC3R) [inline core/lucy.go]: Should come over the `Layer` interface, for readability

File: core/utils.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package lucy
+
```
Comment (L04DB4L4NC3R) [inline core/utils.go:2]: Change name of the file from `utils` to something more descriptive

File: errors/errors.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package lucyErr
+
+import "errors"
+
```
Comment (L04DB4L4NC3R) [inline errors/errors.go:17]: Give errors codes and then map them to type `Error`.

File: core/utils.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	vtype := reflect.TypeOf(v)
+	vvalue := reflect.ValueOf(v)
+
+	if vtype.Kind() != reflect.Struct {
+		vtype = reflect.TypeOf(v).Elem()
+		vvalue = reflect.ValueOf(v).Elem()
+	}
+
+	tagMap := make(map[string]interface{})
+
```
Comment (L04DB4L4NC3R) [inline core/utils.go:20]: `Explanation`: What happens if v is not a struct

File: core/query.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	queries *[]Query
+}
+
+func (q *QueryQueue) Init() {
+	queries := make([]Query, 0)
+	q.queries = &queries
+}
+
+func (q *QueryQueue) Push(query Query) {
+	*q.queries = append(*q.queries, query)
```
Comment (L04DB4L4NC3R) [inline core/query.go]: Potentially dangerous dereference here

File: core/query_engine.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
+package lucy
+
+import (
+	lucyErr "Neo4j-OGM/errors"
+)
+
+type QueryCradle struct {
```
Comment (L04DB4L4NC3R) [inline core/query_engine.go:12]: Comments needed here for explanation

File: neo4j/neo4j.go [hunk 8]
```diff
@@ -11,6 +13,44 @@ type Neo4jRuntime struct {
...
+		"(?:CALL)|" +
+		"(?:YIELD)|" +
+		"(?:USE)|" +
+		"(?:DROP)|" +
+		"(?:START)|" +
+		"(?:STOP)" +
+		")\\s(.*?)")
+)
+
+func (n *Neo4jRuntime) CheckForInjection(expStr string) bool {
```
Comment (L04DB4L4NC3R) [inline neo4j/neo4j.go]: Also return the severity of injection. For example, if an injection can potentially wipe the data.

File: core/query_engine.go [hunk 9]
```diff
@@ -96,7 +103,13 @@ func (q *QueryEngine) Sync() error {
 			if _, ok := cradle.deps[Where]; !ok {
 				return lucyErr.QueryDependencyNotSatisfied
 			}
-			cradle.Exps.Push(qr.Params.(string))
+			param := qr.Params.(string)
+
+			if q.Runtime.CheckForInjection(param) {
+				return lucyErr.QueryInjectionDetected
```
Comment (L04DB4L4NC3R) [inline core/query_engine.go]: Error struct should return details like the severity of injection. For this purpose use custom errors rather than simply assigning variables to `errors.New()`.

Comment (L04DB4L4NC3R) [reply core/query_engine.go]: This will give us much more control on the data of errors that we can serve

File: core/query_engine.go [hunk 10]
```diff
@@ -112,7 +125,14 @@ func (q *QueryEngine) Sync() error {
 				if _, ok := q.cradle.deps[Where]; !ok {
 					return lucyErr.QueryDependencyNotSatisfied
 				}
-				cradle.Exps.Push(qr.Params.(string))
+
+				param := qr.Params.(string)
+
+				if q.Runtime.CheckForInjection(param) {
```
Comment (L04DB4L4NC3R) [inline core/query_engine.go]: Add an explicit flag for disabling `CheckForInjection` at the top level. Keep it enabled by default though

File: core/domain.go [hunk 11]
```diff
@@ -3,19 +3,18 @@ package lucy
 type DomainType uint
```
Comment (L04DB4L4NC3R) [inline core/domain.go]: Again, using enums `iota` would reduce the effort of writing incremental numbers yourself

File: README.md [hunk 12]
```diff
@@ -1,46 +1,38 @@
...
-	<h2 align="center"> < Insert Project Title Here > </h2>
-	<h4 align="center"> < Insert Project Description Here > <h4>
+	<h2 align="center">lucy</h2>
+	<h4 align="center">An easy to use OGM for Neo4j<h4>
 </p>
 
 ---
-[![DOCS](https://img.shields.io/badge/Documentation-see%20docs-green?style=flat-square&logo=appveyor)](INSERT_LINK_FOR_DOCS_HERE) 
-  [![UI ](https://img.shields.io/badge/User%20Interface-Link%20to%20UI-orange?style=flat-square&logo=appveyor)](INSERT_UI_LINK_HERE)
+[![DOCS](https://img.shields.io/badge/Documentation-see%20docs-green?style=flat-square&logo=appveyor)](INSERT_LINK_FOR_DOCS_HERE)
```
Comment (L04DB4L4NC3R) [inline README.md]: Remove this for now

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,9 @@
+.phony: build
+build:
+	PKG_CONFIG_PATH=/usr/local/share/pkgconfig
```
Comment (L04DB4L4NC3R) [inline Makefile:3]: Why is this needed?

**Review by L04DB4L4NC3R — CHANGES_REQUESTED** [review]

File: Makefile (hunk 13 above)
Comment (supercmmetry) [reply Makefile:3]: The go compiler uses pkg-config to link the go binary to static C libs. Hence, the path to pkg-config files must set as an environment variable.

File: core/utils.go (hunk 5 above)
Comment (supercmmetry) [reply core/utils.go:20]: If v is not a struct, it maybe a map, slice or a pointer to a struct. In that case, the value at the base address of v is assumed to be the struct. Then it is marshalled further. In any case if there is any unwanted behaviour in marshalling, the tagMap will remain empty or the program will go into panic. This is an internal function and will not be used by the user directly.

Comment (supercmmetry) [reply core/utils.go:20]: Necessary refactoring will be done.

File: core/utils.go (hunk 3 above)
Comment (supercmmetry) [reply core/utils.go:2]: utils.go contains utility functions used for different purposes. It is better to keep it as it is.

File: core/query.go (hunk 6 above)
Comment (supercmmetry) [reply core/query.go]: The Init() function in the struct, initializes q.queries appropriately. There is no chance of a nil pointer dereference if Init() is called first.

File: core/domain.go [hunk 14]
```diff
@@ -3,18 +3,18 @@ package lucy
...
-	Unknown      DomainType = 7
-	And          DomainType = 8
-	Or           DomainType = 9
-	MiscNodeName DomainType = 10 // Made specifically for dialects
-	Model        DomainType = 11
-	AndStr       DomainType = 12
-	OrStr        DomainType = 13
-	WhereStr     DomainType = 14
+	Where        DomainType = iota
+	Relation     DomainType = iota
```
Comment (L04DB4L4NC3R) [inline core/domain.go]: You only need to write `iota` for the first one. Rest of them are enumerated accordingly

File: errors/errors.go [hunk 15]
```diff
@@ -10,9 +10,9 @@ const (
 )
 
 const (
-	NoSeverity     = iota
-	LowSeverity    = iota
-	HighSeverity   = iota
+	NoSeverity   = iota
+	LowSeverity  = iota
```
Comment (L04DB4L4NC3R) [inline errors/errors.go:14]: You only need to write one `iota`

**Review by L04DB4L4NC3R — APPROVED** [review]

**Comment by L04DB4L4NC3R** [comment]
Change the `iota` comment in the next PR. Rest is fine. Merging
